src/audits/DuplicateId.js

```diff
@@ -25,15 +25,36 @@ axs.AuditRules.addRule({
     heading: 'Any ID referred to via an IDREF must be unique in the DOM',
     url: 'https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_html_02',
     severity: axs.constants.Severity.SEVERE,
-    relevantElementMatcher: function(element) {
-        if (element.hasAttribute('id')) {
-            var referrers = axs.utils.getIdReferrers(element);
-            return referrers.some(function(referrer) {
-                return !axs.utils.isElementOrAncestorHidden(referrer);
+    opt_requires: {
+        domWalk: true,
+        idRefs: true
+    },
+    /**
+     * @this {axs.AuditRule}
+     */
+    relevantElementMatcher: function(element, flags) {
+        if (flags.idrefs.length && !flags.hidden) {
+            this.relatedElements.push({
+                element: element,
+                flags: flags
             });
         }
+        if (element.hasAttribute('id')) {
+            return true;
+        }
         return false;
     },
+    /**
+     * @this {axs.AuditRule}
+     */
+    isRelevant: function(element, flags) {
```

Comment:
This is pretty hard to follow.

It seems like a bunch of infrastructure in this change is simply to make this work. I think it's definitely worthwhile, but it'll be much easier to review if you can split it into two or three changes - potentially one with just the formatting fixes (thanks for that btw), then one with the main performance optimisation (the single pass for collectMatchingElements, and related changes), and then one for this extra optimisation.

I know from experience how annoying this is to do, but git makes it tractable, and it'll make the review a lot easier. Feel free to ping me on hangouts if you need any help with the mechanics - not sure how much of this git surgery you've had to do in the past, but I've definitely done my fair share and happy to share what I've learnt!